pkg/ruler/ruler.go

```diff
@@ -188,66 +162,50 @@ func NewRuler(cfg Config, engine *promql.Engine, queryable storage.Queryable, d
 // Stop stops the Ruler.
 // Each function of the ruler is terminated before leaving the ring
 func (r *Ruler) Stop() {
+	close(r.done)
+	<-r.terminated
+
 	r.notifiersMtx.Lock()
 	for _, n := range r.notifiers {
 		n.stop()
 	}
 	r.notifiersMtx.Unlock()
 
-	level.Info(util.Logger).Log("msg", "shutting down rules scheduler")
-	r.scheduler.Stop()
-
-	level.Info(util.Logger).Log("msg", "waiting for workers to finish")
-	r.workerWG.Wait()
-
 	if r.cfg.EnableSharding {
 		level.Info(util.Logger).Log("msg", "attempting shutdown lifecycle")
 		r.lifecycler.Shutdown()
 		level.Info(util.Logger).Log("msg", "shutting down the ring")
 		r.ring.Stop()
 	}
-}
-
-func (r *Ruler) newGroup(userID string, groupName string, rls []rules.Rule) (*group, error) {
-	appendable := &appendableAppender{pusher: r.pusher}
-	notifier, err := r.getOrCreateNotifier(userID)
-	if err != nil {
-		return nil, err
-	}
-	opts := &rules.ManagerOptions{
-		Appendable:  appendable,
-		QueryFunc:   rules.EngineQueryFunc(r.engine, r.queryable),
-		Context:     context.Background(),
-		ExternalURL: r.alertURL,
-		NotifyFunc:  sendAlerts(notifier, r.alertURL.String()),
-		Logger:      util.Logger,
-		Metrics:     ruleMetrics,
-	}
-	return newGroup(groupName, rls, appendable, opts), nil
-}
 
-func (r *Ruler) removeUser(userID string) error {
-	r.notifiersMtx.Lock()
-	defer r.notifiersMtx.Unlock()
-
-	if n, ok := r.notifiers[userID]; ok {
-		n.stop()
+	level.Info(util.Logger).Log("msg", "stopping user managers")
+	wg := sync.WaitGroup{}
+	r.userManagerMtx.Lock()
+	for user, manager := range r.userManagers {
+		level.Debug(util.Logger).Log("msg", "shutting down user  manager", "user", user)
+		go func(manager *promRules.Manager, user string) {
+			wg.Add(1)
```

Comment:
is this a race condition?  should `wg.Add(1)` be outside the go routine so we are guaranteed it runs before `wg.Wait()`?

Reply:
👍